manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+functions.
+
+\paragraph{Why tail-modulo-cons?} Other program transformations, in
+particular a transformation to continuation-passing style (CPS), can
+make all functions tail-recursive, instead of targeting only a small
+fragment. Some reasons to provide builtin support for the less-general
+tail-mod-cons are as follows:
+\begin{itemize}
+\item The tail-mod-cons transformation preserves the performance of
+  the original, non-tail-recursive version, while
+  a continuation-passing-style transformation incurs a measurable
+  constant-factor overhead.
+\item The tail-mod-cons transformation cannot be expressed as
+  a source-to-source transformation of OCaml programs, as it relies on
+  mutable state in type-unsafe ways. In contrast,
+  continuation-passing-style versions can be written by hand, possibly
+  using a convenient monadic notation.
+\end{itemize}
+
+\subsection{ss:disambiguation}{Disambiguation}
+
+It may happen that several arguments of a constructor are recursive
+calls to a tail-modulo-cons function. The transformation can only turn
+one of these calls into a tail call. It will not make a choice
+implicitly, but ask the user to disambiguate explicitly.
+
+Consider for example this type of syntactic expressions (assuming some
+pre-existing type "var" of expression variables):
+\begin{verbatim}
+type var (* some pre-existing type of variables *)
+
+type exp =
+  | Var of var
+  | Let of binding * exp
+and binding = var * exp
+\end{verbatim}
+
+Consider a "map" function on variables. The direct definition has two
+recursive calls inside arguments of the "Let" constructor, so it gets
+rejected as ambiguous.
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, map_vars f def), map_vars f body)
+
+Error: [@tail_mod_cons]: this constructor application may be TMC-transformed
+       in several different ways. Please disambiguate by adding an explicit
+       [@tailcall] attribute to the call that should be made tail-recursive,
+       or a [@tailcall false] attribute on calls that should not be
+       transformed.
+Line 5, characters 15-29:   This call could be annotated.
+Line 5, characters 32-47:   This call could be annotated.
+\end{verbatim}
+
+To disambiguate, the user should add a "[\@tailcall]" attribute to the
+recursive call that they want to see in tail position:
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, map_vars f def), (map_vars[@tailcall]) f body)
+\end{verbatim}
+Beware that the resulting function is \emph{not} tail-recursive, the
+recursive call on "def" will consume stack space. However, expression
+trees tend to unbalanced on the right (lots of "Let" in sequence,
+rather than nested inside each other), so putting the call on "body"
+in tail position is an interesting improvement over the naive
```

Comment:
I would write `na\"\ive`, but I'm not sure if that's a UK/US thing!

Reply:
I don't know what to do here so I will keep it as-is for now.

(Fun fact: the StackOverflow question ['naive' or 'naïve'](https://english.stackexchange.com/questions/12747/is-it-spelt-na%C3%AFve-or-naive) was closed as a duplicate of the question ['whereäs' as an alternative spelling of 'whereas'](https://english.stackexchange.com/questions/5603/where%c3%a4s-as-an-alternative-spelling-of-whereas).)